src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GaussianMixtureModel.java

```diff
@@ -212,7 +212,11 @@ public Double evaluateDatumInOneDimension( final VariantDatum datum, final int i
         final double[] pVarInGaussianLog10 = new double[gaussians.size()];
         int gaussianIndex = 0;
         for( final MultivariateGaussian gaussian : gaussians ) {
-            pVarInGaussianLog10[gaussianIndex++] = gaussian.pMixtureLog10 + MathUtils.normalDistributionLog10(gaussian.mu[iii], gaussian.sigma.get(iii, iii), datum.annotations[iii]);
+            pVarInGaussianLog10[gaussianIndex] = gaussian.pMixtureLog10;
+            if (gaussian.pMixtureLog10 != Double.NEGATIVE_INFINITY) {
+                pVarInGaussianLog10[gaussianIndex] += MathUtils.normalDistributionLog10(gaussian.mu[iii], gaussian.sigma.get(iii, iii), datum.annotations[iii]);
```

Comment:
It's not clear to me why this special case is needed.  If it's negative infinity, then adding the `normalDistributionLog10` to it yields negative infinity.  You avoid calculating it, of course, but presumably this is an edge case and the optimization doesn't matter.

My other guess is that `sigma` or `mu` could be NaNs somehow?  If so, why is this connected to zero weight?

Reply:
Yeah, sigma and mu are in fact NaNs.  I'm not entirely sure where that happens in the code, but I can look.

Reply:
I think this typically happens when the effective number (i.e., the sum of responsibilities) for the component goes to zero; if so, you can just add an epsilon to the denominators in the updates for the means and covariances to avoid this.  Perhaps see my notebook from an ancient MIA primer for some pointers.

Any plans to update this implementation?  See a few of my objections in https://github.com/broadinstitute/gatk/issues/2062.

Reply:
I think I understand what's going on (that bus number is on the move now).  When a Gaussian has no data assigned the M step `MultivariateGaussian.maximizeGaussian` ends up with zero matrices for `pVarSigma` and `wishart`, and thus the resulting `sigma` equals the `empiricalSigma`, which is just the whole-data (unclustered) covariance matrix.

Since, one of the annotations is constant and has zero variance, this `empiricalSigma` is degenerate.  Since the clusters with data assigned have non-zero `wishart` and `pVarSigma`, only the empty cluster has a problem.

I'm pretty sure that this is wrong, because variational Bayes should regularize the tendency of an empty cluster to take on a degenerate value -- that's the role of the prior.  We could fix it by reconciling the code with Chapter 10 of Bishop (at the cost of several days of dragon-slaying), or we could just re-randomly initialize sigma for an empty cluster, but I would like to the fix the problem at its source so that we avoid creating a pathological Gaussian int he first place.

Reply:
I understand the basic idea of the Bishop approach of dealing with the singularity (mentioned at the end of 10.2.1), but I don't have any confidence in being able to work out the new closed-form solution.  Maybe I can get some help offline?

Reply:
@samuelklee would you prefer a refactored, generalized Java implementation over, for example, a scikit-learn implementation?  You noted your desires at a time when we didn't have any capacity for integrating Python libraries in GATK.

Reply:
Definitely think it would be worthwhile to ultimately extract a Java method/class that calls out to sklearn or some other well tested implementation.  If done right, this could allow easy comparison against other methods available there.  Downsides are of course introducing the python dependency, etc., which might not be something we want to do for a more venerable tool like VQSR.  An intermediate option would be to see if there's a good implementation available in a Java library now.

In any case, I think refactoring to separate out the Bayesian GMM implementation (which I believe I concluded was just standard Bishop, albeit obfuscated by different naming conventions, when I originally wrote those comments---could be wrong, as bus numbers are not only affected by paternity leave but also by time!) from the VQSR-specific machinery is definitely worthwhile.  Not sure how this stacks against other priorities, though.  And I hope I can claim the right to rescind my original offer to be the one to do the work---would be happy to advise, though!

Reply:
I have a number of reasons that I agree it's just standard Bishop.  To fully generalize, I think all that has to happen is to uncouple the actual data from the rest of the `VariantDatum` object, which I can probably do in a not unreasonable amount of time.  I think I modified that class in the past.  I might also be able to fully uncouple the model training from the inference, which I did in a crufty way in the past.  And improve on the memory requirements...

Reply:
I'm not yet convinced that our implementation is faithful to the variational Bayes GMM.  Bishop 10.6 factorizes down to a joint distribution on mu and sigma, but it doesn't split any further, for one thing.  And even if it did, our GMM has a point estimate of sigma for each cluster, which is not the same as a VB M step that computes a posterior.  Finally, I couldn't find anything in Bishop that resembles our `empiricalSigma`, which is the covariance of all the data.

Note that one might worry that Bishop's joint posterior on mu and sigma -- a Gauss-Wishart distribution -- would be intractable for VQSR because the predictive densities could not be computed, but this is not the case.  As Bishop shows, this is just a Student's t.

This is not to say that I think we *should* go after this dragon, but I think it's worth settling whether our implementation is actually a true VB GMM.

Reply:
Don't `empiricalMu` and `empiricalSigma` determine the respective priors?  From a glance, it looks like they are initialized to the origin and some magic number times the identity matrix, respectively, and then never updated.  You might be right about the implementation not being exactly faithful, but I think I concluded that it at least attempted it.

I think that the difficulty of maintaining or improving this code is because 1) the intended implementation is not clearly documented and 2) the code is not easily parsed as the corresponding mathematical expressions.  This is further compounded by the need for the relatively trivial extraction and refactoring mentioned previously.

Reply:
The point estimate you mention (i.e., the one computed by `evaluateFinalModelParameters`) seems to be somewhat arbitrarily chosen, but I think that the updates for the variational parameters calculated in `maximizeGaussian` do seem to follow 10.58 and 10.60-63.

Reply:
Just noting for future reference that I went digging into the original GATK3 code (the commit history over there seems corrupted...not sure if something went wrong at some point, perhaps during a public/protected split?).  It looks like `empiricalMu` and `empiricalSigma` were indeed originally initialized to the mean and covariance of the data (although the code for the latter was commented out), although I'm not sure what the justification for this was.  In https://github.com/broadgsa/gatk/commit/3224bbe7505827f7781ff780023f41e0605fa5be these were changed to represent the quantities for the Bishop prior, but the variable names were not changed.